---
title: "Programming final project"
author: "Cory B Gargas"
date: "May 1, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(reticulate)
```

Scripts mentioned in this project journal that are present in the deliverables folder appear in __BOLD__.  

### __Sequence assignment, filtering, alignment via SSU-ALIGN__ 
  
Using the ttol_rrna dataset created by [AJA](https://www.youtube.com/watch?v=JzrDs_Vaho4&list=PL8a8cutYP7fodkZK1G6kK00rVwhMCy4_F), I created a script to assemble ssu-align commands for each of the extracted transcriptome .fa (unaligned fasta) files.  
  
The file containing this code is named __"ssu-align-bash-script-assembly.sh"__ and contains the code blocked below:  
<pre><code>#! /bin/bash

echo '#! /bin/bash' > ssu-align_ttol-cmds.sh  
echo '#generates .sh file of ssu-align commands for the ttol_rrna data' >> ssu-align_ttol-cmds.sh 

for file in /home/cory/projects/bacteria-diatom-16s/data/tol_rrna/*.fa  
 
do   
 
echo ssu-align -l 250 $file $(basename "${file%.fa}")";" >> ssu-align_ttol-cmds.sh  
        #-l 250 tells ssu-align to filter out sequences below 250 nucleotides  
done

chmod u+x ssu-align_ttol-cmds.sh 
		#makes ssu-align_ttol-cmds.sh executable w/o having to chmod it in terminal  
</code></pre>

I ran the above script using the follow command in the terminal:  
<pre><code>./ssu-align_ttol-cmds.sh > [yr-mo-day]_stdout.txt
</code></pre>  
The file [yr-mo-day]_stdout.txt then contains the stdout for each command of ssu-align run by "ssu-align_ttol-cmds.sh"  
  
This file ends up producing the usual SSU-ALIGN output for each .fa file inside of a folder named after the input .fa file.  

I then used the following codes to concatenate the bacterial and archaeal sequences for each domain:   
<pre><code>cat */*.archaea.fa > archaea_250bp.fa;    
cat */*.bacteria.fa > bacteria_250bp.fa;  
</code></pre> 
  
I then ran the following commands to find the total number of sequences for Archaea and Bacteria in the dataset after using SSU-ALIGN to filter sequences <250bp.  

<pre><code>#! /bin/bash
grep -c ">" 2019-02-26_output_250bp/archaea_250bp.fa >> compare-analyses.txt;
grep -c ">" 2019-02-26_output_250bp/bacteria_250bp.fa >> compare-analyses.txt;
</code></pre>
  
I then used the following [bioawk](https://github.com/lh3/bioawk) commands to sort these files by name and sequence length into tab-delimited text files.  
  
<pre> 
bioawk -c fastx '{ print $name, length($seq) }' < archaea-250bp.fa > archaea-250bp-lengths.txt  
#commands for bacteria were the same as for archaea.  
</code></pre>   

The *-lengths.txt files were imported into r code (In the raw .rmd file for the project journal) to create the following distribution plots:    
  
```{r , echo=FALSE, warning=FALSE, message=FALSE}
#completed on 2019-03-06

library(ggplot2)
library(reshape)
library(plyr)
theme_set(theme_classic())

#reading in the tab delimited .txt files
a.250.len <- read.delim("~/projects/bacteria-diatom-16s/docs/proj-journal/archaea-250bp-lengths.txt", header=FALSE)
b.250.len <- read.delim("~/projects/bacteria-diatom-16s/docs/proj-journal/bacteria-250bp-lengths.txt", header=FALSE)


#prepare the data by removing names and combinging lists
colnames(a.250.len)[colnames(a.250.len)=="V2"] <- "a250bp"
colnames(b.250.len)[colnames(b.250.len)=="V2"] <- "b250bp"

a.250.len = subset(a.250.len, select = -c(V1))
b.250.len = subset(b.250.len, select = -c(V1))

mcat.a.250 <- melt(a.250.len)
mcat.b.250 <- melt(b.250.len)

final <- rbind(mcat.a.250,mcat.b.250)

#generates means for our plots
final.mu <- ddply(final, "variable", summarise, grp.mean=mean(value))

#labels for each plot
hist.lab <- list(
  'a250bp'="Archaea, total=57",
  'b250bp'="Bacteria, total=11865"
  )

labeller <- function(variable,value){
  return(hist.lab[value])
}

###############
#histogram version#
###############


ggplot(final, aes(x=value, fill=variable)) +
  geom_histogram(aes(x=value), binwidth = 15, alpha=0.9) +
  scale_fill_brewer(palette="Dark2") + 
  theme_minimal() +
  facet_wrap(variable ~ ., scales = "free", labeller = labeller) +
  scale_x_continuous(name="Seqeunce length", breaks = c(0,250,500,750,1000,1250,1500,1750), limits = c(0, 1750)) +
  geom_vline(data=final.mu, aes(xintercept=grp.mean, color=variable), linetype="dashed") +
  scale_color_brewer(palette="Dark2")


########################
#histogram with density overlay#
########################

#ggplot(final, aes(x=value,fill=variable)) + 
#  geom_histogram(aes(y=..density..), binwidth=15)+
#  geom_density(alpha=0) +
#  scale_fill_brewer(palette="Dark2") + 
#  theme_minimal() +
#  facet_wrap(variable ~ ., scales="free") +
#  scale_y_continuous(name="density",  limits = c(0,0.008)) +
#  scale_x_continuous(name="Seqeunce length", breaks = c(0,250,500,750,1000,1250,1500,1750), limits = c(0,1750)) +
#  geom_vline(data=mu, aes(xintercept=grp.mean, color=variable), linetype="dashed") +
#  scale_color_brewer(palette="Dark2") 
```

  
Table 1. Results of SSU-ALIGN filtering with a minimum sequence length of 250bp.  
  
domain | All | 250bp  
--- | --- | --- 
Archaea | 67 | 57   
Bacteria | 18385 | 11865  

  
### __Reference phylogeny__  
  
The following options were available to be used as reference alignments and trees:  

+ [Greengenes](http://greengenes.secondgenome.com/downloads)  
+ [SILVA/“All-Species Living Tree Project” (LTP)](http://www.arb-silva.de/projects/living-tree)  
+ [EzBioCloud](https://www.ezbiocloud.net/)  
+ [RDP](https://rdp.cme.msu.edu/)  
+ [NCBI](https://www.ncbi.nlm.nih.gov/)  
+ [OTT](https://tree.opentreeoflife.org/about/taxonomy-version/ott3.0)

All have pros and cons:  

Table 2. Overview of five taxonomic classifications. Adapted and updated from: [SILVA, RDP, Greengenes, NCBI and OTT — how do these taxonomies compare? (2017) Monika Balvočiūtė and Daniel H. Huson](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-017-3501-4)  
  
Taxonomy | Type | No. of nodes | Lowest rank | Latest release  
--------- | --------- | --------- | --------- | ---------  
Greengenes | Automatic | 3,093 | Species | May 2013  
RDP | Semi | 6,128 | Genus | Sep 2016  
SILVA | Manual | 12,117 | Genus | Jun 2018  
EzBioCloud | Automatic? | ~63,376 | Species | Feb 2019  
NCBI | Manual | 1,522,150 | Species | All day, erry day  
OTT | Automatic | 2,627,066 | Species | Feb 2017  
  
Decided to use the SILVA RTP database as it is the most tractable in size and is updated frequently.  
  
### __Fixing LTP reference files__

There were several issues with the LTP reference files that needed to be remediated before they could be used.  
__1.__  The fasta file headers had a mix of spaces and tabs, so I needed to replace both of these with underscores.  
__2.__  The fasta file sequences had a mix of periods and hyphens to represent missing sites, so I needed to repalce periods with hyphens.  
__3.__  The sequence names in the fasta file and the tip names in the tree file were different, making it impossible to actually map those sequences onto their     reference tree.  
__4.__  The acutal fasta headers and tip labels were all gratuitously long and contained unnecessary information, so I wanted to clean those up to just contain the following: accession#_taxonomy(genus, species, family, etc).  

To fix these issues I created a python script  __(LTP_fasta_and_tree_renamer.py)__ to correct these issues and then output corrected fasta and tree files. This script also produces a CSV file containing all the extra info from the sequence headers.  
  
### __OTU clustering via [CD-hit](https://github.com/weizhongli/cdhit)__   

Used a sequence identity threshold (-c) of 0.97 as that's the common cutoff in traditional metabarcoding studies.  
  
The following is the code used to execute the analyses.  
<pre><code>
cd-hit-est -i Archaea_250bp.fa -o archaea-250bp.cdhit.fa -c 0.97 -n 10 -d 0 -M 8000 -T 4  
</code></pre> 
  
resulting files are stored in the directory cdhit-97.  
  
I used grep -c ">" on the resulting .fa files to find how many redundant sequences were removed.  

Used bioawk once again to get sequence lengths for the resulting.fa files:  
  
<pre><code>
bioawk -c fastx '{ print $name, length($seq) }' < archaea-250bp.cdhit.fa > cat-250bp-cd-len.txt  
</code></pre> 

Using the same r code as previously, I plotted the following distributions for my data.   

```{r , echo=FALSE, warning=FALSE, message=FALSE}

library(ggplot2)
library(reshape)
library(plyr)
theme_set(theme_classic())

#reading in the tab delimited .txt files
a97.250.len <- read.delim("/home/cory/projects/bacteria-diatom-16s/data/cdhit/cdhit-c97/archaea-250bp-cd-len.txt", header=FALSE)
b97.250.len <- read.delim("/home/cory/projects/bacteria-diatom-16s/data/cdhit/cdhit-c97/bacteria-250bp-cd-len.txt", header=FALSE)


#prepare the data by removing names and combinging lists
colnames(a97.250.len)[colnames(a97.250.len)=="V2"] <- "a250bp"
colnames(b97.250.len)[colnames(b97.250.len)=="V2"] <- "b250bp"

#remove the first column
a97.250.len = subset(a97.250.len, select = -c(V1))
b97.250.len = subset(b97.250.len, select = -c(V1))



#melt them to create the variable column indicating how they were processed
a97.250.len <- melt(a97.250.len)
b97.250.len <- melt(b97.250.len)

#bind the dataframes for each domain/ssu-align run in each -c value of CD-hit that we've run
c97.final <- rbind(a97.250.len, b97.250.len)


#generates means for our plots
c97.mu <- ddply(c97.final, "variable", summarise, grp.mean=mean(value))

##########
#c97 density#
##########
c97.hist.lab <- list(
  'a250bp'="Archaea, total=19",
  'b250bp'="Bacteria, total=5604"
  )

c97.labeller <- function(variable,value){
  return(c97.hist.lab[value])
}

#ggplot(c97.final, aes(x=value, fill=variable)) +
#  geom_density(alpha=0.8) +
#  scale_fill_brewer(palette="Dark2") + 
#  theme_minimal() +
#  facet_wrap(variable ~ ., scales = "free", labeller = c97.labeller) +
#  scale_x_continuous(name="Seqeunce length", breaks = c(0,250,500,750,1000,1250,1500,1750), limits = c(0, 1750)) +
#  geom_vline(data=c97.mu, aes(xintercept=grp.mean, color=variable), linetype="dashed") +
#  scale_color_brewer(palette="Dark2")

############
#c97 histogram #
############

ggplot(c97.final, aes(x=value, fill=variable)) +
  geom_histogram(aes(x=value), binwidth = 15, alpha=0.9) +
  scale_fill_brewer(palette="Dark2") + 
  theme_minimal() +
  facet_wrap(variable ~ ., scales = "free", labeller = c97.labeller) +
  scale_x_continuous(name="c97 seqeunce length", breaks = c(0,250,500,750,1000,1250,1500,1750), limits = c(0, 1750)) +
  geom_vline(data=c97.mu, aes(xintercept=grp.mean, color=variable), linetype="dashed") +
  scale_color_brewer(palette="Dark2")
```

Table 1. Data attenuation after CD-hit OTU clustering at 97 ident.  
  
domain | All | 250bp | CD-hit.c97 
--- | --- | --- | --- | 
Archaea | 67 | 57 | 19 |  
Bacteria | 18385 | 11865 | 5604 |   

### __Aligning archaeal and bacterial sequences__  

Before RAxML phylogenetic inferences, must align and mask all of the individual fasta files from the transcriptomes.  
To accomplish this I used the SSU-ALIGN package from before, then masked them using the default masking (-d) as implemetned in the ssu-mask command.  

Accomplished this by using the following commands.  
  
<pre><code>
ssu-align -f archaea-250bp.cdhit.fa archaea-out  
ssu-mask -d --afa archaea-out/
</code></pre> 

ssu-align found that 5 of the 19 archaea sequences were misidentified bacterial sequences.  
These were then concatenated with the bacteria cdhit .fa file and ssualign was performed for the bacterial sequences.  
  
Table 1. Data attenuation after 2nd SSU-align and concatenating misidentifed sequences.  
  
domain | All | SSU-ALIGN.250bp | CD-hit.c97 | SSU-ALIGN.alignment |
--- | --- | --- | --- | --- |
Archaea | 67 | 57 | 19 | 14 |  
Bacteria | 18385 | 11865 | 5604 | 5609 |   


### __Profile alignment using MUSCLE__  
  
Now that I've aligned and masked the sequences using ssu-align, I need to create a profile alignment of our individual bacterial and archaeal alignments against the Silva LTP alignment using MUSCLE. 

Because of the large amount of processing power this alignment type needs I used the AHPCC Razor cluster to conduct my profile alignment.  

I used the following PBS script to submit my MUSCLE commands to the cluster and align our archaea .afa against the LTP_132 .afa
<pre><code>
#!/bin/bash

#PBS -N muscle
#PBS -q mem512GB64core
#PBS -j oe
#PBS -m abe
#PBS -M cbgargas@uark.edu
#PBS -o muscle.$PBS_JOBID
#PBS -l nodes=1:ppn=64
#PBS -l walltime=00:72:00:00

cd $PBS_O_WORKDIR

module purge
module load muscle/3.8.31

muscle -profile -in1 LTPs132_SSU_aligned.fasta -in2 archaea-out.archaea.mask.afa -out archaea-LTP132_aligned.afa 
</code></pre> 


The above pbs script was generated using the python script __pbs-script-gen.py.__
  

# __*FIN*__  


